### PR TITLE
shell support: make `which spack` output intelligible

### PR DESCRIPTION
**Pull request by tgamblin**
Zsh and newer versions of bash have a builtin `which` function that will show you if a command is actually an alias or a function. For functions, the entire function is printed, and our `spack()` function is quite long. Instead of printing out all that, make the `spack()` function a wrapper around `_spack_shell_wrapper()`, and include some no-ops in the definition so that users can see where it was created and where Spack is installed.

Here's what the new output looks like in zsh:

```console
$ which spack
spack () {
	: this is a shell function from: /Users/gamblin2/src/spack/share/spack/setup-env.sh
	: the real spack script is here: /Users/gamblin2/src/spack/bin/spack
	_spack_shell_wrapper "$@"
	return $?
}
```

Note that `:` is a no-op in Bourne shell; it just discards anything after it on the line. We use it here to embed paths in the function definition (as comments are stripped).

### CODE REVIEW

**Comment by tgamblin** [comment]
@sameershende FYI

**Review by adamjstewart — APPROVED** [review]

**Comment by alalazo** [comment]
Closed/reopen to get the `mpich` hotfix that was merged for Github's CI

**Review by michaelkuhn — APPROVED** [review]
works for me

**Review by alalazo — APPROVED** [review]
Works for me too on MacOS Catalina

File: share/spack/setup-env.sh [hunk 1]
```diff
@@ -315,6 +311,21 @@ if ! _spack_fn_exists use && ! _spack_fn_exists module; then
 	need_module="yes"
 fi;
 
+# Define the spack shell function with some informative no-ops, so when users
+# run `which spack`, they see the path to spack and where the function is from.
+eval "spack() {
```
Comment (tldahlgren) [inline share/spack/setup-env.sh:316]: LGTM.  

If this a *real* lack of coverage issue, please add a test for the `eval` here to get the patch coverage up.

**Review by tldahlgren — DISMISSED** [review]
Please address the patch coverage failure IF it is real.

**Comment by tgamblin** [comment]
Coverage issue here is not real, so I'm going to go ahead and merge.

**Pull request merged**